Move middleware to lib

sonalkr132 opened the pull request:
Fixes in rails 6 update: A copy of Hostess has been removed from the module tree but is
still active! (ArgumentError) in development.rb

`app` folder is autoloaded in development and causes this error. It is
adviced to not put middleware (not reloadable) in app folder.
> You should not put middlewares or any code that should not be
reloadable in the app folder. (rafaelfranca, [rails/rails#25525](https://github.com/rails/rails/issues/25525#issuecomment-229071607))

require_relative is used as `lib` is not present in load path during
initialization.

~~Interestingly, our `lib` is also in `autoload_paths` as we are using
 `config.eager_load_paths << Rails.root.join('lib')` in `application.rb`. 
I guess that is not same autoload as app folder(?)~~

simi commented:
Can you just move it to lib directly? It will be autoloaded then IMHO. Otherwise you can add `lib/middleware` to load paths. There will be no need for any additional require then AFAIK.

sonalkr132 commented:
> Can you just move it to lib directly? It will be autoloaded 

`A copy of Hostess has been removed from the module tree but is still active! (ArgumentError)` is being raised because `Hostess` is being autoloaded and [`RubygemFs` referenced in it](https://github.com/rubygems/rubygems.org/blob/master/app/middleware/hostess.rb#L3), is not.
Moving middleware out of autoload path is the only way I have been able to fix this.

We can go back to keeping middleware in the autloaded path (and remove `require_relative`) when we are using `config.load_defaults 6.0`. I guess this is related to Zeitwerk changes. I haven't reported this rails as I haven't been able to write git bisect file for this exception (how does one modify `Rails::Application.autoload_path` in [a file like this](https://gist.github.com/prakashmurthy/d03093d8f7f135ad91e5567308a61428#file-decimal_rounding_error-rb)?) and I have already been suggested to move middleware code out of the autoload path on the [older tickets](https://github.com/rails/rails/issues/25525#issuecomment-229071607).

sonalkr132 commented:
Hi, merging this. I have removed changes from production.rb and we can go back to keeping all middlewares in the app folder when we use `config.load_defaults 6.0`.
I hope this addresses your concerns.
